test/test_c10d.py

```diff
@@ -367,5 +430,90 @@ def allreduce(tensors, op):
+        return F.softmax(x, dim=1)
+
+
+class DistributedDataParallelTest(MultiProcessTestCase):
+
+    @property
+    def size(self):
+        return 2
+
+    def _test_ddp_with_process_group(self, process_group):
+        gpus = gpus_for_rank(self.size)[self.rank]
+        model = Net()
+        ddp_model = distributed_c10d._DistributedDataParallelC10d(
+            copy.deepcopy(model).cuda(gpus[0]),
+            process_group,
+            device_ids=gpus)
+        model.cuda(gpus[0])
+
+        local_batch_size = len(gpus)
+        global_batch_size = self.size * local_batch_size
+        criterion = nn.MSELoss()
+        input = torch.randn(global_batch_size, 2).cuda(gpus[0])
+        target = torch.randn(global_batch_size, 4).cuda(gpus[0])
+
+        def step_model(model, input, target, criterion):
+            model.train()
+            output = model(input)
+            loss = criterion(output, target)
+            loss.backward()
+
+        def update_parameters(model):
+            for param in model.parameters():
+                param.data -= param.grad
+                param.grad = None
+
+        # check two model parameters over 2 iterations
+        for _ in range(2):
+            # single cpu/gpu training
+            step_model(model, input, target, criterion)
+
+            # DDP training, DDP scatters subsets of input_cpu to nodes/GPUs
+            step_model(ddp_model,
+                       input[self.rank * local_batch_size: (self.rank + 1) * local_batch_size],
+                       target[self.rank * local_batch_size: (self.rank + 1) * local_batch_size],
+                       criterion)
+
+            # Update weights and run a second iteration to shake out errors
+            update_parameters(model)
+            update_parameters(ddp_model)
+            assert len(list(model.parameters())) == len(list(ddp_model.parameters()))
+            for i, j in zip(model.parameters(), ddp_model.parameters()):
+                self.assertEqual(i, j)
+
+            # Shuffle the input so that DDP input is different
+            input = input[torch.randperm(global_batch_size)]
+
+    @skip_if_not_multigpu
+    def test_gloo_backend(self):
+        store = c10d.TCPStore('localhost', self.port, self.rank == 0)
+        options = c10d.ProcessGroupGloo.Options()
+        options.devices = [c10d.ProcessGroupGloo.create_tcp_device(interface="lo")]
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.size, options)
+        self._test_ddp_with_process_group(process_group)
+
+    @skip_if_not_multigpu
+    @skip_if_not_nccl
+    def test_nccl_backend(self):
+        store = c10d.TCPStore('localhost', self.port, self.rank == 0)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.size)
+        self._test_ddp_with_process_group(process_group)
+
 if __name__ == '__main__':
+    assert not torch.cuda._initialized, "test_distributed must not have initialized CUDA context on main process"
```

Comment:
Yep if I ever accidentally initialized cuda on the main process, the initialization would fail on all other processes. For example I couldn't use `torch.cuda.device_count()` on the main process. This line tells us whenever we do initialize CUDA on the main process so we can fix it